cohorts/load.py

```diff
@@ -23,131 +23,311 @@
+        for cache_name in cache_names.values():
+            cache_dir = path.join(self.cache_dir, cache_name)
+            if path.exists(cache_dir):
+                cache_subdirs = set(listdir(cache_dir))
+                cache_int_subdirs = set([int(name) for name in cache_subdirs])
+                if len(cache_subdirs) != len(cache_int_subdirs):
+                    bad_caches.append(cache_name)
+
+        if len(bad_caches) > 0:
+            raise ValueError("Caches %s have duplicate int/str directories" %
+                             str(bad_caches))
+
+    def as_dataframe(self, join_with=None, join_how=None):
+        join_dataframes = []
+
+        # Use join_with if specified, otherwise fall back to what is defined in the class
+        join_with = first_not_none_param([join_with, self.join_with], default=[])
+        if type(join_with) == str:
+            join_with = [join_with]
+
+        # Use join_how if specified, otherwise fall back to what is defined in the class
+        join_how = first_not_none_param([join_how, self.join_how], default="outer")
+
+        df = pd.DataFrame()
+        df["patient_id"] = pd.Series([patient.id for patient in self])
+        for clinical_col in ["benefit", "os", "pfs", "deceased",
+                             "progressed", "progressed_or_deceased"]:
+            df[clinical_col] = pd.Series([getattr(patient, clinical_col) for patient in self])
+
+        additional_data_all_patients = defaultdict(list)
+        for patient in self:
+            if patient.additional_data is not None:
+                for key, value in patient.additional_data.items():
+                    additional_data_all_patients[key].append(value)
+
+        if len(additional_data_all_patients) > 0:
+            df = df.merge(pd.DataFrame(additional_data_all_patients), on="patient_id", how="left")
+
+        for df_loader in join_with:
+            old_len_df = len(df)
+            df = df.merge(
+                df_loader.load_dataframe(),
+                left_on="patient_id",
+                right_on=df_loader.join_on,
+                how=join_how)
+            print("%s join with %s: %d to %d rows" % (
+                join_how,
+                df_loader.name,
+                old_len_df,
+                len(df)))
+        return df
+
+    def load_from_cache(self, cache_name, patient_id, file_name):
         if not self.cache_results:
             return None
 
         cache_dir = path.join(self.cache_dir, cache_name)
-        sample_cache_dir = path.join(cache_dir, str(sample_id))
-        cache_file = path.join(sample_cache_dir, file_name)
+        patient_cache_dir = path.join(cache_dir, str(patient_id))
+        cache_file = path.join(patient_cache_dir, file_name)
 
         if not path.exists(cache_file):
             return None
 
         if path.splitext(cache_file)[1] == ".csv":
-            return pd.read_csv(cache_file)
+            return pd.read_csv(cache_file, dtype={"patient_id": object})
```

Comment:
Maybe  special-case (and can be handled later) but `chromosome` always kills me this way. If there are no X, Y mutations or "chr" prefix pandas will load them as ints and kill any joins.